[LVI] Generalize mask not equal conditions handling

antoniofrighetto opened the pull request:
Extend `V & Mask != 0` for non-zero constants, when retrieving constraint value information from a non-equality comparison.

Proof: https://alive2.llvm.org/ce/z/dc5BeT.
Motivating example: https://github.com/gcc-mirror/gcc/blob/master/gcc/testsuite/gcc.dg/tree-ssa/vrp76.c

goldsteinn commented on llvm/test/Transforms/CorrelatedValuePropagation/icmp.ll:
```diff
@@ -1475,3 +1474,37 @@ entry:
+
+end:
+  ret void
+}
```
Can you please split the test to a seperate commit so we can see the diff from your code changes?

nikic reviewed (changes requested):
I'm pretty sure this is wrong. You can only simply invert the range if it is an *exact* one.

If you take a satisfying range for pred, then the inverse is an allowed range for !pred, not a satisfying range.

nikic commented:
To give an example, let's say C=0 (ignoring the existing special case). Then the range for EQ will be `[0, ~Mask]`. The range for NE you would compute is `[~Mask, MAX]`, which is (very) incorrect.

antoniofrighetto commented:
I realize I've completely mistook inverse implementation. The original idea was to merge ranges `[SIGN_MIN, R.Lower)` and `[R.Upper, SIGN_MAX + 1)`, where `R` is the range obtained from EQ, then erroneously fell back to inverse; but even so, I now realize the original idea is wrong too. Is it still possible to save something out of it? Perhaps checking if we fall within `makeExactICmpRegion`?

antoniofrighetto commented:
@nikic Can I assume that `R = GetRangeFromKnownBits()` is always a satisfying range for pred EQ?  For the motivating example (range: `[-16, -8)`), shouldn't I expect that `makeAllowedICmpRegion(!pred, R.inverse()) == makeSatisfyingICmpRegion(!pred, R.inverse())`? Sorry if I'm still off track :(

nikic commented:
I don't really get your question. It should be possible to do *something* here, but I'd expect it to be some generalization of the existing `& Mask != 0` case.

antoniofrighetto commented:
> To give an example, let's say C=0 (ignoring the existing special case). Then the range for EQ will be `[0, ~Mask]`. The range for NE you would compute is `[~Mask + 1, MAX]`, which is (very) incorrect.

Revisited this better, yet I'm still unsure. If C=0, the range for EQ is indeed `[0, ~Mask]`, so the inverse is `[~Mask + 1, MAX]`, that is precisely the one you obtain in the current existing case, which should be correct. Could you please show another example where inverse leads to improper result? AFAICT we should be wrong only when `(Val & Mask) == C` is unsat, in which case we ensure to be conservative.

nikic commented:
> > To give an example, let's say C=0 (ignoring the existing special case). Then the range for EQ will be `[0, ~Mask]`. The range for NE you would compute is `[~Mask + 1, MAX]`, which is (very) incorrect.
> 
> Revisited this better, yet I'm still unsure. If C=0, the range for EQ is indeed `[0, ~Mask]`, so the inverse is `[~Mask + 1, MAX]`, that is precisely the one you obtain in the current existing case, which should be correct. Could you please show another example where inverse leads to improper result? AFAICT we should be wrong only when `(Val & Mask) == C` is unsat, in which case we ensure to be conservative.

No, the existing case returns something different. Let's say you have Mask = 0b0011. Then valid values would be 0b0001, 0b0010 and 0b0011 (but not 0b0000). The existing code returns `[0b0001, MAX]`, not `[0b1101, MAX]`.

antoniofrighetto committed:
[LVI][ConstantRange] Generalize mask not equal conditions handling

Extend `V & Mask != 0` for non-zero constants if satisfiable, when
retrieving constraint value information from a non-equality comparison.

Proof: https://alive2.llvm.org/ce/z/dc5BeT.

Motivating example: https://github.com/gcc-mirror/gcc/blob/master/gcc/testsuite/gcc.dg/tree-ssa/vrp76.c.

antoniofrighetto commented:
Thanks `@nikic`! I think we should be on the path towards getting this on track now.

nikic commented on llvm/test/Analysis/LazyValueAnalysis/lvi-icmp-ne-of-and.ll:
```diff
@@ -0,0 +1,91 @@
+; RUN: opt < %s -passes="jump-threading,print<lazy-value-info>" -disable-output 2>&1 | FileCheck %s
```
I prefer not to have `print<lazy-value-info>` tests. If you want to have better test coverage for this logic, I would recommend adding a `ConstantRange::makeMastNotEqualRange()` or so helper and then test it in ConstantRangeTest -- the big advantage is that you can add an exhaustive test and be certain that the logic is always correct.

antoniofrighetto replied:
Good point! Updated and dropped previous test, thanks.

goldsteinn commented:
Can you generalize your proofs?

nikic commented on llvm/lib/IR/ConstantRange.cpp:
```diff
@@ -364,6 +364,29 @@ ConstantRange ConstantRange::makeExactNoWrapRegion(Instruction::BinaryOps BinOp,
 
+ConstantRange ConstantRange::makeMaskNotEqualRange(const APInt &Mask,
+                                                   const APInt &C) {
+  assert(!Mask.isZero() && "Mask cannot be zero.");
```
I feel like this pre-condition should not exist -- we can handle zero mask inside this function.

nikic commented on llvm/lib/IR/ConstantRange.cpp:
```diff
@@ -364,6 +364,29 @@ ConstantRange ConstantRange::makeExactNoWrapRegion(Instruction::BinaryOps BinOp,
+  if (C.isZero())
+    return ConstantRange::getNonEmpty(
+        APInt::getOneBitSet(BitWidth, TrailingZeroesOfMask),
+        APInt::getZero(BitWidth));
```
Do we need this special case? I think for C=0 the below case is equivalent to this one...

antoniofrighetto replied:
It it is indeed equivalent, but thought it was a bit nice / helpful to have the base case explicit.

nikic replied:
IMHO its best not to have unnecessary special cases...

nikic commented on llvm/unittests/IR/ConstantRangeTest.cpp:
```diff
@@ -2788,4 +2788,31 @@ TEST_F(ConstantRangeTest, isSizeLargerThan) {
+
+  for (unsigned MaskVal = 1; MaskVal < Max; ++MaskVal) {
+    APInt Mask(Bits, MaskVal);
+    for (unsigned CVal = 0; CVal < Max; ++CVal) {
```
Can use EnumerateAPInts here.

nikic commented on llvm/unittests/IR/ConstantRangeTest.cpp:
```diff
@@ -2788,4 +2788,31 @@ TEST_F(ConstantRangeTest, isSizeLargerThan) {
+      // {0, 4, 8, 12}, however we conservatively return [1, 0).
+      auto CR = ConstantRange::makeMaskNotEqualRange(Mask, C);
+      TestRange(CR, Elems, PreferSmallestUnsigned, {}, false);
+      TestRange(CR, Elems, PreferSmallestSigned, {}, false);
```
It doesn't make sense to call this with two different PreferSmallest function if you set CheckOptimality=false (in which case it will be ignored).

nikic commented on llvm/unittests/IR/ConstantRangeTest.cpp:
```diff
@@ -2788,4 +2788,31 @@ TEST_F(ConstantRangeTest, isSizeLargerThan) {
   EXPECT_FALSE(One.isSizeLargerThan(1));
 }
 
+TEST_F(ConstantRangeTest, MakeMaskNotEqualRangeExhaustive) {
```
I'd suggest to also add one or two tests with explicit ranges (as we're not testing optimality).

nikic commented on llvm/unittests/IR/ConstantRangeTest.cpp:
```diff
@@ -2788,4 +2788,31 @@ TEST_F(ConstantRangeTest, isSizeLargerThan) {
+
+      // Do not check for optimality, as levelling off for efficiency. E.g.,
+      // given Mask = 0b0011, C = 0b0000, the optimal range would be FullSet ∖
+      // {0, 4, 8, 12}, however we conservatively return [1, 0).
```
Note that `[1, 0)` *is* an optimal range in this case. Optimal isn't the same as precise, it's just the best we can do with a ConstantRange envelope.

antoniofrighetto replied:
Right, comment rectified.

antoniofrighetto commented:
> Can you generalize your proofs?

Definitely, updated PR w/ generalized proof (https://alive2.llvm.org/ce/z/dc5BeT).

nikic commented on llvm/unittests/IR/ConstantRangeTest.cpp:
```diff
@@ -2788,4 +2788,52 @@ TEST_F(ConstantRangeTest, isSizeLargerThan) {
+      // a possible better range would be [0, 15) when preferring the smallest
+      // unsigned, however we conservatively return [2, 1).
+      TestRange(ConstantRange::makeMaskNotEqualRange(Mask, C), Elems,
+                PreferSmallestUnsigned, {}, false);
```
Just to check, if you used PreferSmallest instead of PreferSmallestUnsigned here, it would still not pass the optimality check, right?

antoniofrighetto replied:
Right, but only due to the fact that we were conservatively returning full-set when `Elems.none()` holded. Now returning empty-set in the only occurring case, so switched to PreferSmallest with optimality set, thanks for noticing it.

nikic commented on llvm/lib/IR/ConstantRange.cpp:
```diff
@@ -364,6 +364,26 @@ ConstantRange ConstantRange::makeExactNoWrapRegion(Instruction::BinaryOps BinOp,
+  // If (Val & Mask) != C, constrained to the non-equality being
+  // satisfiable, then the value must be larger than the lowest set bit of
+  // Mask, offset by constant C.
+  if ((Mask & C) == C)
```
I think it would be slightly more elegant if you do a `if ((Mask & C) != C) return getFull(BitWidth);` check at the start of the function? Then in the Mask.isZero() case you can always return getEmpty().

nikic reviewed (approved):
LGTM
